### ISSUE TITLE
Weird output when getting buffer with XSynth_ChannelGroup_ReadSamples

### ISSUE THREAD

**Issue by KaleidonKep99** [MyBlackMIDIScore/xsynth-clib#2]
I don't know if I made a mistake in the code or not, but all I get is a buzzing noise.
From the audio clip I sent, it almost sounds like the buffer constantly starts from the beginning every time I call the ReadSamples function, adding the new NoteOns on top.

Audio example:

https://github.com/user-attachments/assets/2efcf042-10a9-4b9f-903a-66ccf53d40c5

This is how I parse the buffer from XSynth (I had to remove "* sizeof(float)" next to the count, since keeping that there caused the RAM usage to skyrocket to 64+ GB for some reason):
```cs
public override unsafe int Read(float[] buffer, int offset, int count)
{
    lock (Lock)
    {
        fixed (float* buff = buffer)
        {
            var offsetBuff = buff + offset;

            XSynth.ChannelGroup_ReadSamples(handle, (nint)offsetBuff, (ulong)count);

            return count;
        }
    }
}
```

This is how the original function looks like for BASSMIDI, in case you need a reference:
```cs
public override unsafe int Read(float[] buffer, int offset, int count)
{
    lock (Lock)
    {
        fixed (float* buff = buffer)
        {
            var offsetBuff = buff + offset;
            var len = (count * sizeof(float)) | (WaveFormat.BitsPerSample == 32 ? (int)DataFlags.Float : 0);

            int ret = Bass.ChannelGetData(Handle, (nint)offsetBuff, len);
            if (ret == 0)
            {
                var BE = Bass.LastError;

                if (BE != Errors.Ended)
                    Debug.PrintToConsole(Debug.LogType.Warning, $"{UniqueID} - Data parsing error {BE} with length {len}");
            }

            return ret / 4;
        }
    }
}
```

**Comment by MyBlackMIDIScore** [comment]
> I had to remove "* sizeof(float)" next to the count, since keeping that there caused the RAM usage to skyrocket to 64+ GB for some reason

Yeah the `count` parameter is just the number of audio samples, not the size of the array.

I have a question about this code, because I am not familiar with how BASSMIDI renders audio. Is the `buffer` parameter the output of the entire audio file and the `offset` the index in the list where it is supposed to start rendering from?
If so, the `count` should be the amount of samples that correspond to the last MIDI event sent.
So it should be `count = sample_rate * audio_channel_count * delta_time`.

An example of how this should be done is [this function in Forte](https://github.com/MyBlackMIDIScore/forte/blob/407701de9964775b4ce866b8281e10e4694cb2ca/src/xsynth/midi_pool.rs#L208) where `self.renderer.read_samples(&mut self.output_vec);` should work like `XSynth_ChannelGroup_ReadSamples()`. If you have any questions about this code let me know.

So yeah let me know exactly what the parameters of your function do so I can have a better picture.

**Comment by MyBlackMIDIScore** [comment]
Wait are you casting the type of the buffer to integer here? 
```c#
(nint)offsetBuff
```

Since the `Read` function accepts a float buffer and XSynth renders only in float (f32) type, try removing the cast to see if that does anything.

**Comment by KaleidonKep99** [comment]
> > I had to remove "* sizeof(float)" next to the count, since keeping that there caused the RAM usage to skyrocket to 64+ GB for some reason
> 
> Yeah the `count` parameter is just the number of audio samples, not the size of the array.
> 
> I have a question about this code, because I am not familiar with how BASSMIDI renders audio. Is the `buffer` parameter the output of the entire audio file and the `offset` the index in the list where it is supposed to start rendering from? If so, the `count` should be the amount of samples that correspond to the last MIDI event sent. So it should be `count = sample_rate * audio_channel_count * delta_time`.
> 
> An example of how this should be done is [this function in Forte](https://github.com/MyBlackMIDIScore/forte/blob/407701de9964775b4ce866b8281e10e4694cb2ca/src/xsynth/midi_pool.rs#L208) where `self.renderer.read_samples(&mut self.output_vec);` should work like `XSynth_ChannelGroup_ReadSamples()`. If you have any questions about this code let me know.
> 
> So yeah let me know exactly what the parameters of your function do so I can have a better picture.

Hi, the offset option is currently unused here so it's nothing to worry about. All it really does is make it so you write to the buffer after a specific position in the array. Nothing in OCv2 makes use of it yet so it's all good.

And I should've guessed that ReadSamples work in samples and not bytes... 💀
I'll do some conversion to convert the byte count into samples, then I'll let you know if it works!

> Wait are you casting the type of the buffer to integer here?
> 
> ```cs
> (nint)offsetBuff
> ```
> 
> Since the `Read` function accepts a float buffer and XSynth renders only in float (f32) type, try removing the cast to see if that does anything.

nint is just C#'s way of calling 32/64-bit pointers, it's just like a void*.

**Comment by MyBlackMIDIScore** [comment]
> Hi, the offset option is currently unused here so it's nothing to worry about. All it really does is make it so you write to the buffer after a specific position in the array. Nothing in OCv2 makes use of it yet so it's all good.

Ah alright

> And I should've guessed that ReadSamples work in samples and not bytes, as the name implies... 💀
I'll do some conversion to convert the byte count into samples, then I'll let you know if it works!

Lol no worries, let me know if it worked when you try this

> nint is just C#'s way of calling 32/64-bit pointers, it's just like a void*.

Oooh I see, interesting

**Comment by KaleidonKep99** [comment]
I don't think I understand how this all works. I managed to make it sound *better*, even though saying better in this context is seriously stretching the meaning of the word.

Feeding it the amount of samples I need *plus* the delta just makes the whole app segfault, so I do not think I can do that. From what I gathered so far from the numerous crashes, XSynth seems to expect me to apply an offset to the sample request, but I am not sure how to do that? Adding the delta does not work as mentioned earlier, and there's no other parameter that would allow me to just do "give me *N* samples beginning from *OFF* offset of your internal stream".

https://github.com/user-attachments/assets/7e63085d-eeec-4391-a190-1745b5694177

Here's how BASS's data parsing function works: https://www.un4seen.com/doc/#bass/BASS_ChannelGetData.html

tl;dr is that I just request the amount of bytes I need, and BASS gives them to me. It probably has internal read and write heads, the write head moving forward whenever it receives a MIDI event, and the reading one moving forward whenever BASS_ChannelGetData is called, up until it runs out of data. XSynth always starts from the beginning of the stream, which is exactly what **I don't** want it to do... 🥹

**Comment by KaleidonKep99** [comment]
I completely forgot to paste in the code used for BASSMIDI, to give some context as to how it all works... Oops.

Here's a tl;dr of what the code is supposed to do, removing all the unnecessary parts:
```cs
public void ProcessTheEvents(ISampleWriter output, WaveFormat waveFormat, IEnumerable<MIDIEvent> events) {
    var midiRenderer = new XSynthRenderer();
    float[] buffer = new float[64 * waveFormat.BlockAlign];
    byte[] scratch = new byte[16];

    foreach (var e in events) {
        var eb = e.GetData(scratch);

        long writeTime = (long)(deltaTime * waveFormat.SampleRate);

        // If writeTime ends up being negative, clamp it to 0
        if (writeTime < prevWriteTime)
            writeTime = prevWriteTime;

        var offset = (int)((writeTime - prevWriteTime) * 2);

        // Never EVER go back in time!!!!!!
        if (writeTime > prevWriteTime) // <<<<< EVER!!!!!!!!!
            prevWriteTime = writeTime;

        while (offset > 0)
        {
            bool smallOffset = offset < buffer.Length;
            midiRenderer.Read(buffer, 0, smallOffset ? offset : buffer.Length);
            output.Write(buffer, 0, smallOffset ? offset : buffer.Length);
            offset = smallOffset ? 0 : offset - buffer.Length;
        }

        // do midi processing here...
    }

    if (midiRenderer.SendEndEvent())
    {
        while ((read = midiRenderer.Read(buffer, 0, buffer.Length)) != 0)
            output.Write(buffer, 0, read);
    }
    else
    {
        // There's no true way to guess when the stream ends with XSynth,
        // since it does not return any value from ReadSamples,
        // so I just check if the value of the first byte is 0, signaling no audio
        var fl = 1.0f;

        while (fl != 0.0f)
        {
            midiRenderer.Read(buffer, 0, buffer.Length);
            output.Write(buffer, 0, buffer.Length);

            fl = buffer[0];
        }
    }
}
```

**Comment by MyBlackMIDIScore** [comment]
Upon having a first look at that code block I can't really figure out what might be wrong. 

Would you be able to send me the source code of OC so I can test it out myself?

**Comment by KaleidonKep99** [comment]
> Upon having a first look at that code block I can't really figure out what might be wrong.
> 
> Would you be able to send me the source code of OC so I can test it out myself?

I'll just commit the changes, so you can clone it on your machine.
It is different from the example I sent you, since I had to implement some hacky changes to get anything coherent out of XSynth's output.

https://github.com/KaleidonKep99/OmniConverter/tree/ed982e4748f7784db7e4ada1da01bc83e9e41693

**Comment by MyBlackMIDIScore** [comment]
> I'll just commit the changes, so you can clone it on your machine. It is different from the example I sent you, since I had to implement some hacky changes to get anything coherent out of XSynth's output.
> 
> https://github.com/KaleidonKep99/OmniConverter/tree/ed982e4748f7784db7e4ada1da01bc83e9e41693

Great thanks! I'll have a look when I get back home and I'll let you know

**Comment by MyBlackMIDIScore** [comment]
@KaleidonKep99 I just tried using it and I am getting an `Error: Init` from `ManagedBass` when trying to use this and I am not sure why. I can't figure out what is causing it. Am I doing something wrong?

![image](https://github.com/user-attachments/assets/8cb8d9a9-80f4-4914-a095-48fd5cbaf09c)

**Comment by MyBlackMIDIScore** [comment]
Couldn't get it to work currently, but upon hearing the audio file you sent with headphones I think I know what the issue might be. 

When writing to the provided buffer, XSynth does not replace the data in it, rather it sums the new data with the existing.

This is bad behavior, so I will have to change this in the main project too. But for now, try setting all the values of `buffer` to 0 before each `ReadSamples` call and let me know if that fixes the issue.

**Comment by KaleidonKep99** [comment]
> @KaleidonKep99 I just tried using it and I am getting an `Error: Init` from `ManagedBass` when trying to use this and I am not sure why. I can't figure out what is causing it. Am I doing something wrong?
> 
> ![image](https://private-user-images.githubusercontent.com/31241793/354246759-8cb8d9a9-80f4-4914-a095-48fd5cbaf09c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI1MjcwOTMsIm5iZiI6MTcyMjUyNjc5MywicGF0aCI6Ii8zMTI0MTc5My8zNTQyNDY3NTktOGNiOGQ5YTktODBmNC00OTE0LWEwOTUtNDhmZDVjYmFmMDljLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODAxVDE1Mzk1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJkYTVjZWE3N2Q4NDFmMjdlMDA4MjIxZGY5NGM5Y2VhOThlMmE3Y2YzNjAxOWUzMzNmZmYyZDFlYTg3ODNkY2MmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.4Q86RefuT79O4OVQU4HQzbNWqlSzJ9a47bzX2plGiOk)
> 
> EDIT: I have added all the necessary DLLs, so it shouldn't be the issue. I am running it in a VM, I don't know if this has anything to do with the error
> 
> EDIT2: I also tried it building it on Linux but I get the same error

Yeah that is a bug in BASS, disabling the conversion sounds in the settings fixes the issue.

> Couldn't get it to work currently, but upon hearing the audio file you sent with headphones I think I know what the issue might be.
> 
> When writing to the provided buffer, XSynth does not replace the data in it, rather it sums the new data with the existing.
> 
> This is bad behavior, so I will have to change this in the main project too. But for now, try setting all the values of `buffer` to 0 before each `ReadSamples` call and let me know if that fixes the issue.

So, if XSynth detects that a buffer has data, it just appends more to it? That's really weird. I will do what you said I guess.

**Comment by KaleidonKep99** [comment]
It seems to work now, albeit a bit clicky, but I guess that's because of the really low layer count. All good! Let me know when the fixed ReadSamples is out.

https://github.com/user-attachments/assets/a37071cd-4354-4e9f-bc52-5367b7d94865

Thank you! 🫡

**Comment by MyBlackMIDIScore** [comment]
> Yeah that is a bug in BASS, disabling the conversion sounds in the settings fixes the issue.

Oh lol didn't know that, I'll try it, thanks.

> So, if XSynth detects that a buffer has data, it just appends more to it? That's really weird. I will do what you said I guess.

> Let me know when the fixed ReadSamples is out.

Yeah it's really bad practice but I never bothered to fix it, although I will do it soon because it only creates confusion really. I will ping you here when I do. Glad it worked eventually :)

> albeit a bit clicky

Should be the `fade_out_killing` option for `ChannelGroup` which is false by default for performance reasons, but I'd recommend you to set it to true by default and maybe make an option for it, because it can cause a lot of clicking due to how voices are managed.

**Comment by KaleidonKep99** [comment]
> > So, if XSynth detects that a buffer has data, it just appends more to it? That's really weird. I will do what you said I guess.
> 
> > Let me know when the fixed ReadSamples is out.
> 
> Yeah it's really bad practice but I never bothered to fix it, although I will do it soon because it only creates confusion really. I will ping you here when I do. Glad it worked eventually :)

Yeah I usually just expect the data parsing functions to just overwrite the whole buffer with its own data, that threw me off for a loop I'll tell you much!

> 
> > albeit a bit clicky
> 
> Should be the `fade_out_killing` option for `ChannelGroup` which is false by default for performance reasons, but I'd recommend you to set it to true by default and maybe make an option for it, because it can cause a lot of clicking due to how voices are managed.

Yeah, it was set to *true* for that export, but it didn't work for some reason. 🤷 

I did encounter another problem with the library, I'll make a new issue about it. 😢
